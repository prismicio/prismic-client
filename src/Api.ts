import Predicates from './Predicates';
import { RequestHandler, RequestCallback } from './request';
import { Document } from './documents';
import { ApiCache } from './cache';
import ResolvedApi, { ApiData } from './ResolvedApi';
import HttpClient from './HttpClient';

export interface ApiOptions {
  accessToken?: string;
  requestHandler?: RequestHandler;
  req?: any;
  apiCache?: ApiCache;
  apiDataTTL?: number;
  proxyAgent?: any;
}

export default class PrismicApi {
  url: string;
  options: ApiOptions;
  apiDataTTL: number;
  httpClient: HttpClient;

  constructor(url: string, options?: ApiOptions) {
    this.options = options || {};
    this.url = url;
    if (this.options.accessToken) {
      const accessTokenParam = `access_token=${this.options.accessToken}`;
      this.url += (url.indexOf('?') > -1 ? '&' : '?') + accessTokenParam;
    }
    this.apiDataTTL = this.options.apiDataTTL || 5;
    this.httpClient = new HttpClient(
      this.options.requestHandler,
      this.options.apiCache,
      this.options.proxyAgent,
    );
  }

  /**
   * Fetches data used to construct the api client, from cache if it's
   * present, otherwise from calling the prismic api endpoint (which is
   * then cached).
   */
  get(cb?: RequestCallback<ResolvedApi>): Promise<ResolvedApi> {
    return this.httpClient.cachedRequest<ApiData>(this.url, { ttl: this.apiDataTTL }).then((data) => {
      const resolvedApi = new ResolvedApi(data, this.httpClient, this.options);
      cb && cb(null, resolvedApi);
      return resolvedApi;
    }).catch((error) => {
      cb && cb(error);
      throw error;
    });
  }
}
