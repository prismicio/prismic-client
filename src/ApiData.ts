export interface QuickRoutes {
  url: string;
  lastUpdate: number;
  enabled: boolean;
}

export interface ApiData {
  refs: Ref[];
  bookmarks: { [key: string]: string };
  types: { [key: string]: string };
  tags: string[];
  forms: { [key: string]: Form };
  quickRoutes: QuickRoutes;
  experiments: any;
  oauth_initiate: string;
  oauth_token: string;
  version: string;
  licence: string;
}
