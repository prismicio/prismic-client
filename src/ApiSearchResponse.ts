import { Document } from "./documents";

export default interface ApiSearchResponse<T = any> {
  page: number;
  results_per_page: number;
  results_size: number;
  total_results_size: number;
  total_pages: number;
  next_page: string;
  prev_page: string;
  results: Document<T>[];
}
