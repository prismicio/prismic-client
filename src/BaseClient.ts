import { type LimitFunction, pLimit } from "./lib/pLimit"

import { PrismicError } from "./errors/PrismicError"

/**
 * The default delay used with APIs not providing rate limit headers.
 *
 * @internal
 */
export const UNKNOWN_RATE_LIMIT_DELAY = 1500

/**
 * A universal API to make network requests. A subset of the `fetch()` API.
 *
 * {@link https://developer.mozilla.org/en-US/docs/Web/API/fetch}
 */
export type FetchLike = (
	input: string,
	init?: RequestInitLike,
) => Promise<ResponseLike>

/**
 * An object that allows you to abort a `fetch()` request if needed via an
 * `AbortController` object
 *
 * {@link https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal}
 */
// `any` is used often here to ensure this type is universally valid among
// different AbortSignal implementations. The types of each property are not
// important to validate since it is blindly passed to a given `fetch()`
// function.
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type AbortSignalLike = any

/**
 * A subset of RequestInit properties to configure a `fetch()` request.
 */
// Only options relevant to the client are included. Extending from the full
// RequestInit would cause issues, such as accepting Header objects.
//
// An interface is used to allow other libraries to augment the type with
// environment-specific types.
export interface RequestInitLike extends Pick<RequestInit, "cache"> {
	/**
	 * The HTTP method to use for the request.
	 */
	method?: string

	/**
	 * The request body to send with the request.
	 */
	// We want to keep the body type as compatible as possible, so
	// we only declare the type we need and accept anything else.
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	body?: any | FormData | string

	/**
	 * An object literal to set the `fetch()` request's headers.
	 */
	headers?: Record<string, string>

	/**
	 * An AbortSignal to set the `fetch()` request's signal.
	 *
	 * See:
	 * [https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal)
	 */
	// NOTE: `AbortSignalLike` is `any`! It is left as `AbortSignalLike`
	// for backwards compatibility (the type is exported) and to signal to
	// other readers that this should be an AbortSignal-like object.
	signal?: AbortSignalLike
}

/**
 * The minimum required properties from Response.
 */
export interface ResponseLike {
	ok: boolean
	status: number
	headers: HeadersLike
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	json(): Promise<any>
	text(): Promise<string>
	blob(): Promise<Blob>
}

/**
 * The minimum required properties from Headers.
 */
export interface HeadersLike {
	get(name: string): string | null
}

/**
 * The minimum required properties to treat as an HTTP Request for automatic
 * Prismic preview support.
 */
export type HttpRequestLike =
	| /**
	 * Web API Request
	 *
	 * @see http://developer.mozilla.org/en-US/docs/Web/API/Request
	 */
	{
			headers?: {
				get(name: string): string | null
			}
			url?: string
	  }

	/**
	 * Express-style Request
	 */
	| {
			headers?: {
				cookie?: string
			}
			query?: Record<string, unknown>
	  }

/**
 * Configuration for clients that determine how APIs are queried.
 */
export type BaseClientConfig = {
	/**
	 * The function used to make network requests to the Prismic REST API. In
	 * environments where a global `fetch` function does not exist, such as
	 * Node.js, this function must be provided.
	 */
	fetch?: FetchLike

	/**
	 * Options provided to the client's `fetch()` on all network requests. These
	 * options will be merged with internally required options. They can also be
	 * overriden on a per-query basis using the query's `fetchOptions` parameter.
	 */
	fetchOptions?: RequestInitLike
}

/**
 * Parameters for any client method that use `fetch()`.
 */
export type FetchParams = {
	/**
	 * Options provided to the client's `fetch()` on all network requests. These
	 * options will be merged with internally required options. They can also be
	 * overriden on a per-query basis using the query's `fetchOptions` parameter.
	 */
	fetchOptions?: RequestInitLike

	/**
	 * An `AbortSignal` provided by an `AbortController`. This allows the network
	 * request to be cancelled if necessary.
	 *
	 * @deprecated Move the `signal` parameter into `fetchOptions.signal`:
	 *
	 * @see \<https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal\>
	 */
	signal?: AbortSignalLike
}

/**
 * The result of a `fetch()` job.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type FetchJobResult<TJSON = any> = {
	status: number
	headers: HeadersLike
	json: TJSON
	text?: string
}

export abstract class BaseClient {
	/**
	 * The function used to make network requests to the Prismic REST API. In
	 * environments where a global `fetch` function does not exist, such as
	 * Node.js, this function must be provided.
	 */
	fetchFn: FetchLike

	fetchOptions?: RequestInitLike

	/**
	 * Active queued `fetch()` jobs keyed by URL and AbortSignal (if it exists).
	 */
	private queuedFetchJobs: Record<string, LimitFunction> = {}

	/**
	 * Active deduped `fetch()` jobs keyed by URL and AbortSignal (if it exists).
	 */
	private dedupedFetchJobs: Record<
		string,
		Map<AbortSignalLike | undefined, Promise<FetchJobResult>>
	> = {}

	constructor(options: BaseClientConfig) {
		this.fetchOptions = options.fetchOptions

		if (typeof options.fetch === "function") {
			this.fetchFn = options.fetch
		} else if (typeof globalThis.fetch === "function") {
			this.fetchFn = globalThis.fetch as FetchLike
		} else {
			throw new PrismicError(
				"A valid fetch implementation was not provided. In environments where fetch is not available (including Node.js), a fetch implementation must be provided via a polyfill or the `fetch` option.",
				undefined,
				undefined,
			)
		}

		// If the global fetch function is used, we must bind it to the global scope.
		if (this.fetchFn === globalThis.fetch) {
			this.fetchFn = this.fetchFn.bind(globalThis)
		}
	}

	protected async fetch(
		url: string,
		params: FetchParams = {},
	): Promise<FetchJobResult> {
		const requestInit: RequestInitLike = {
			...this.fetchOptions,
			...params.fetchOptions,
			headers: {
				...this.fetchOptions?.headers,
				...params.fetchOptions?.headers,
			},
			signal:
				params.fetchOptions?.signal ||
				params.signal ||
				this.fetchOptions?.signal,
		}

		// Request with a `body` are throttled, others are deduped.
		if (params.fetchOptions?.body) {
			return this.queueFetch(url, requestInit)
		} else {
			return this.dedupeFetch(url, requestInit)
		}
	}

	private queueFetch(
		url: string,
		requestInit: RequestInitLike = {},
	): Promise<FetchJobResult> {
		// Rate limiting is done per hostname.
		const hostname = new URL(url).hostname

		if (!this.queuedFetchJobs[hostname]) {
			this.queuedFetchJobs[hostname] = pLimit({
				interval: UNKNOWN_RATE_LIMIT_DELAY,
			})
		}

		return this.queuedFetchJobs[hostname](() =>
			this.createFetchJob(url, requestInit),
		)
	}

	private dedupeFetch(
		url: string,
		requestInit: RequestInitLike = {},
	): Promise<FetchJobResult> {
		let job: Promise<FetchJobResult>

		// `fetchJobs` is keyed twice: first by the URL and again by is
		// signal, if one exists.
		//
		// Using two keys allows us to reuse fetch requests for
		// equivalent URLs, but eject when we detect unique signals.
		if (
			this.dedupedFetchJobs[url] &&
			this.dedupedFetchJobs[url].has(requestInit.signal)
		) {
			// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
			job = this.dedupedFetchJobs[url].get(requestInit.signal)!
		} else {
			this.dedupedFetchJobs[url] = this.dedupedFetchJobs[url] || new Map()

			job = this.createFetchJob(url, requestInit).finally(() => {
				this.dedupedFetchJobs[url]?.delete(requestInit.signal)

				if (this.dedupedFetchJobs[url]?.size === 0) {
					delete this.dedupedFetchJobs[url]
				}
			})

			this.dedupedFetchJobs[url].set(requestInit.signal, job)
		}

		return job
	}

	private createFetchJob(
		url: string,
		requestInit: RequestInitLike = {},
	): Promise<FetchJobResult> {
		return this.fetchFn(url, requestInit).then(async (res) => {
			// We can assume Prismic REST API responses
			// will have a `application/json`
			// Content Type. If not, this will
			// throw, signaling an invalid
			// response.
			// eslint-disable-next-line @typescript-eslint/no-explicit-any
			let json: any = undefined
			let text: string | undefined = undefined
			if (res.ok) {
				try {
					json = await res.json()
				} catch {
					// noop
				}
			} else {
				try {
					text = await res.text()
					json = JSON.parse(text)
				} catch {
					// noop
				}
			}

			return {
				status: res.status,
				headers: res.headers,
				json,
				text,
			}
		})
	}
}
