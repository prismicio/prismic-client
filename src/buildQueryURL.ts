import { castArray } from "./lib/castArray";

import { ValueOf, Ordering, Route } from "./types";

/**
 * Parameters for the Prismic REST API V2.
 *
 * {@link https://prismic.io/docs/technologies/introduction-to-the-content-query-api}
 */
export interface QueryParams {
	/**
	 * The secure token for accessing the API (only needed if your repository is
	 * set to private).
	 *
	 * {@link https://user-guides.prismic.io/en/articles/1036153-generating-an-access-token}
	 */
	accessToken?: string;

	/**
	 * The `pageSize` parameter defines the maximum number of documents that the
	 * API will return for your query.
	 *
	 * {@link https://prismic.io/docs/technologies/search-parameters-reference-rest-api#pagesize}
	 */
	pageSize?: number;

	/**
	 * The `page` parameter defines the pagination for the result of your query.
	 *
	 * {@link https://prismic.io/docs/technologies/search-parameters-reference-rest-api#page}
	 */
	page?: number;

	/**
	 * The `after` parameter can be used along with the orderings option. It will
	 * remove all the documents except for those after the specified document in the list.
	 *
	 * {@link https://prismic.io/docs/technologies/search-parameters-reference-rest-api#after}
	 */
	after?: string;

	/**
	 * The `fetch` parameter is used to make queries faster by only retrieving the
	 * specified field(s).
	 *
	 * {@link https://prismic.io/docs/technologies/search-parameters-reference-rest-api#fetch}
	 */
	fetch?: string | string[];

	/**
	 * The `fetchLinks` parameter allows you to retrieve a specific content field
	 * from a linked document and add it to the document response object.
	 *
	 * {@link https://prismic.io/docs/technologies/search-parameters-reference-rest-api#fetchlinks}
	 */
	fetchLinks?: string | string[];

	/**
	 * The `graphQuery` parameter allows you to specify which fields to retrieve
	 * and what content to retrieve from Linked Documents / Content Relationships.
	 *
	 * {@link https://prismic.io/docs/technologies/graphquery-rest-api}
	 */
	graphQuery?: string;

	/**
	 * The `lang` option defines the language code for the results of your query.
	 *
	 * {@link https://prismic.io/docs/technologies/search-parameters-reference-rest-api#lang}
	 */
	lang?: string;

	/**
	 * The `orderings` parameter orders the results by the specified field(s). You
	 * can specify as many fields as you want.
	 *
	 * {@link https://prismic.io/docs/technologies/search-parameters-reference-rest-api#orderings}
	 */
	orderings?: Ordering | string | (Ordering | string)[];

	/**
	 * The `routes` option allows you to define how a document's `url` field is resolved.
	 *
	 * {@link https://prismic.io/docs/core-concepts/link-resolver-route-resolver#route-resolver}
	 */
	routes?: Route | string | (Route | string)[];

	/**
	 * The `brokenRoute` option allows you to define the route populated in the
	 * `url` property for broken Link or Content Relationship fields. A broken
	 * link is a Link or Content Relationship field whose linked document has been
	 * unpublished or deleted.
	 *
	 * {@link https://prismic.io/docs/core-concepts/link-resolver-route-resolver#route-resolver}
	 */
	brokenRoute?: string;
}

/**
 * Arguments for `buildQueryURL` to construct a Query URL.
 */
type BuildQueryURLParams = {
	/**
	 * Ref used to query documents.
	 *
	 * {@link https://prismic.io/docs/technologies/introduction-to-the-content-query-api#prismic-api-ref}
	 */
	ref: string;

	/**
	 * Ref used to populate Integration Fields with the latest content.
	 *
	 * {@link https://prismic.io/docs/core-concepts/integration-fields}
	 */
	integrationFieldsRef?: string;

	/**
	 * One or more predicates to filter documents for the query.
	 *
	 * {@link https://prismic.io/docs/technologies/query-predicates-reference-rest-api}
	 */
	predicates?: string | string[];
};

/**
 * Parameters in this map have been renamed from the official Prismic REST API
 * V2 specification for better developer ergonomics.
 *
 * These parameters are renamed to their mapped value.
 */
const RENAMED_PARAMS = {
	accessToken: "access_token",
} as const;

/**
 * A valid parameter name for the Prismic REST API V2.
 */
type ValidParamName =
	| Exclude<
			keyof QueryParams,
			keyof typeof RENAMED_PARAMS | keyof BuildQueryURLParams
	  >
	| ValueOf<typeof RENAMED_PARAMS>;

/**
 * Converts an Ordering to a string that is compatible with Prismic's REST API.
 * If the value provided is already a string, no conversion is performed.
 *
 * @param ordering - Ordering to convert.
 *
 * @returns String representation of the Ordering.
 */
const castOrderingToString = (ordering: Ordering | string): string =>
	typeof ordering === "string"
		? ordering
		: [
				ordering.field,
				ordering.direction === "desc" ? ordering.direction : undefined,
		  ]
				.filter(Boolean)
				.join(" ");

export type BuildQueryURLArgs = QueryParams & BuildQueryURLParams;

/**
 * Build a Prismic REST API V2 URL to request documents from a repository. The
 * paginated response for this URL includes documents matching the parameters.
 *
 * A ref is required to make a request. Request the `endpoint` URL to retrieve a
 * list of available refs.
 *
 * Type the JSON response with `Query`.
 *
 * {@link https://prismic.io/docs/technologies/introduction-to-the-content-query-api#prismic-api-ref}
 * {@link https://prismic.io/docs/technologies/query-predicates-reference-rest-api}
 *
 * @param endpoint - URL to the repository's REST API V2.
 * @param args - Arguments to filter and scope the query.
 *
 * @returns URL that can be used to request documents from the repository.
 */
export const buildQueryURL = (
	endpoint: string,
	args: BuildQueryURLArgs,
): string => {
	const { predicates, ...params } = args;

	const url = new URL(`documents/search`, `${endpoint}/`);

	if (predicates) {
		for (const predicate of castArray(predicates)) {
			url.searchParams.append("q", `[${predicate}]`);
		}
	}

	// Iterate over each parameter and add it to the URL. In some cases, the
	// parameter value needs to be transformed to fit the REST API.
	for (const k in params) {
		const name = (RENAMED_PARAMS[k as keyof typeof RENAMED_PARAMS] ??
			k) as ValidParamName;

		let value = params[k as keyof typeof params];

		if (name === "orderings") {
			const scopedValue = params[name];

			if (scopedValue != null) {
				const v = castArray(scopedValue)
					.map((ordering) => castOrderingToString(ordering))
					.join(",");

				value = `[${v}]`;
			}
		} else if (name === "routes") {
			if (typeof params[name] === "object") {
				value = JSON.stringify(castArray(params[name]));
			}
		}

		if (value != null) {
			url.searchParams.set(name, castArray(value).join(","));
		}
	}

	return url.toString();
};
