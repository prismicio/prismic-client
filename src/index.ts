// Primary library API.
import { getRepositoryEndpoint } from "./getRepositoryEndpoint";
export { getRepositoryEndpoint };
/**
 * @deprecated Renamed to `getRepositoryEndpoint`.
 */
// TODO: Remove in v3.
export const getEndpoint = getRepositoryEndpoint;
export { getRepositoryName } from "./getRepositoryName";
export { getGraphQLEndpoint } from "./getGraphQLEndpoint";
export { isRepositoryName } from "./isRepositoryName";
export { isRepositoryEndpoint } from "./isRepositoryEndpoint";
export { buildQueryURL } from "./buildQueryURL";
export { createClient, Client } from "./client";

// Predicates API.
import { predicate } from "./predicate";
export { predicate };
/**
 * @deprecated Renamed to `predicate` (without an "s").
 */
// TODO: Remove in v3.
export const predicates = predicate;
/**
 * @deprecated Renamed to `predicate` (lowercase and without an "s").
 */
// TODO: Remove in v3.
export const Predicates = predicate;

// Custom errors used by Client.
export { PrismicError } from "./PrismicError";
export { ForbiddenError } from "./ForbiddenError";
export { ParsingError } from "./ParsingError";
export { NotFoundError } from "./NotFoundError";

// A collection of well-known cookie names shared between Prismic libraries and systems.
export * as cookie from "./cookie";

// General types used throughout the project. These are made public to allow users to better type their projects.
export type { CreateClient, ClientConfig } from "./client";
export type { QueryParams, BuildQueryURLArgs } from "./buildQueryURL";
export type {
	AbortSignalLike,
	FetchLike,
	HttpRequestLike,
	Ordering,
	RequestInitLike,
	ResponseLike,
	Route,
} from "./types";
