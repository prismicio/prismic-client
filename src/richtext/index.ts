export { asTree } from "./asTree";
export { asText } from "./asText";

export { serialize } from "./serialize";
export { wrapMapSerializer } from "./wrapMapSerializer";
export { composeSerializers } from "./composeSerializers";

export { RichTextNodeType as Element } from "../types/value/richText";

export type {
	RichTextFunctionSerializer,
	RichTextMapSerializer,
	RichTextMapSerializerFunction,
} from "./types";
