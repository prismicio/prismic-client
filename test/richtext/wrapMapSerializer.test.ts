import { expect, it } from "vitest"

import { createRichTextFixtures } from "../__testutils__/createRichTextFixtures"

import { htmlRichTextMapSerializer } from "../__fixtures__/htmlRichTextMapSerializer"

import { serialize, wrapMapSerializer } from "../../src/richtext"

it("serializes a rich text field value using a given serializer map", () => {
	const richTextFixtures = createRichTextFixtures()

	const serializer = wrapMapSerializer(htmlRichTextMapSerializer)
	const serialization = serialize(richTextFixtures.en, serializer)

	expect(serialization).toMatchSnapshot()
})

it("returns `undefined` when serializer does not handle given tag", () => {
	const richTextFixtures = createRichTextFixtures()

	const serializer = wrapMapSerializer({})
	const serialization = serialize(richTextFixtures.en, serializer)

	expect(serialization.every((element) => element === undefined)).toBe(true)
})
