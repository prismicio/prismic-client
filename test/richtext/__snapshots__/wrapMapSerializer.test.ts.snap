// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`serializes a rich text field value using a given serializer map 1`] = `
[
  "<h1>Lorem ipsum dolor sit amet.</h1>",
  "<h2>Lorem ipsum dolor sit amet.</h2>",
  "<h3>Lorem ipsum dolor sit amet.</h3>",
  "<h4>Lorem ipsum dolor sit amet.</h4>",
  "<h5>Lorem ipsum dolor sit amet.</h5>",
  "<h6>Lorem ipsum dolor sit amet.</h6>",
  "<p>Lorem ipsum dolor sit amet.</p>",
  "<p></p>",
  "<h1>Lorem ipsum dolor <strong>sit amet</strong>, consectetur <em>adipiscing elit</em>, sed do eiusmod tempor incididunt ut labore et dolore <span class=\\"label\\">magna aliqua</span>. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea <a href=\\"https://google.com\\" target=\\"undefined\\">commodo consequat</a>. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat <a href=\\"linkResolver(XvoFFREAAM0WGBng)\\">nulla pariatur</a>. Excepteur sint occaecat cupidatat <a href=\\"https://images.prismic.io/200629-sms-hoy/f0a757f6-770d-4eb8-a08b-f1727f1a58e4_guilherme-romano-KI2KaOeT670-unsplash.jpg?auto=compress,format\\">non proident</a>, sunt in culpa qui officia deserunt mollit anim id est laborum.</h1>",
  "<p>Lorem ipsum dolor <strong>sit amet</strong>, consectetur <em>adipiscing elit</em>, sed do eiusmod tempor incididunt ut labore et dolore <span class=\\"testLabel\\">magna aliqua</span>. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea <a href=\\"https://google.com\\" target=\\"undefined\\">commodo consequat</a>. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat <a href=\\"linkResolver(XvoFFREAAM0WGBng)\\">nulla pariatur</a>. Excepteur sint occaecat cupidatat <a href=\\"https://images.prismic.io/200629-sms-hoy/f0a757f6-770d-4eb8-a08b-f1727f1a58e4_guilherme-romano-KI2KaOeT670-unsplash.jpg?auto=compress,format\\">non proident</a>, sunt in culpa qui officia deserunt mollit anim id est laborum.</p>",
  "<ol><li>Lorem ipsum dolor <strong>sit amet</strong>, consectetur <em>adipiscing elit</em>, sed do eiusmod tempor incididunt ut labore et dolore <span class=\\"label\\">magna aliqua</span>. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea <a href=\\"https://google.com\\" target=\\"undefined\\">commodo consequat</a>. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat <a href=\\"linkResolver(XvoFFREAAM0WGBng)\\">nulla pariatur</a>. Excepteur sint occaecat cupidatat <a href=\\"https://images.prismic.io/200629-sms-hoy/f0a757f6-770d-4eb8-a08b-f1727f1a58e4_guilherme-romano-KI2KaOeT670-unsplash.jpg?auto=compress,format\\">non proident</a>, sunt in culpa qui officia deserunt mollit anim id est laborum.</li><li>Lorem ipsum dolor sit amet.</li><li>Lorem ipsum dolor sit amet.</li></ol>",
  "<ul><li>Lorem ipsum dolor <strong>sit amet</strong>, consectetur <em>adipiscing elit</em>, sed do eiusmod tempor incididunt ut labore et dolore <span class=\\"label\\">magna aliqua</span>. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea <a href=\\"https://google.com\\" target=\\"undefined\\">commodo consequat</a>. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat <a href=\\"linkResolver(XvoFFREAAM0WGBng)\\">nulla pariatur</a>. Excepteur sint occaecat cupidatat <a href=\\"https://images.prismic.io/200629-sms-hoy/f0a757f6-770d-4eb8-a08b-f1727f1a58e4_guilherme-romano-KI2KaOeT670-unsplash.jpg?auto=compress,format\\">non proident</a>, sunt in culpa qui officia deserunt mollit anim id est laborum.</li><li>Lorem ipsum dolor sit amet.</li><li>Lorem ipsum dolor sit amet.</li></ul>",
  "<p>Lorem ipsum dolor <strong>sit amet</strong>, consectetur <em>adipiscing elit</em>, sed do eiusmod tempor incididunt ut labore et dolore <span class=\\"label\\">magna aliqua</span>. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea <a href=\\"https://google.com\\" target=\\"undefined\\">commodo consequat</a>. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat <a href=\\"linkResolver(XvoFFREAAM0WGBng)\\">nulla pariatur</a>. Excepteur sint occaecat cupidatat <a href=\\"https://images.prismic.io/200629-sms-hoy/f0a757f6-770d-4eb8-a08b-f1727f1a58e4_guilherme-romano-KI2KaOeT670-unsplash.jpg?auto=compress,format\\">non proident</a>, sunt in culpa qui officia deserunt mollit anim id est laborum.</p>",
  "<p>Lorem ipsum dolor <strong>sit amet</strong>, consectetur <em>adipiscing elit</em>, sed do eiusmod tempor incididunt ut labore et dolore <span class=\\"label\\">magna aliqua</span>. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea <a href=\\"https://google.com\\" target=\\"undefined\\">commodo consequat</a>. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat <a href=\\"linkResolver(XvoFFREAAM0WGBng)\\">nulla pariatur</a>. Excepteur sint occaecat cupidatat <a href=\\"https://images.prismic.io/200629-sms-hoy/f0a757f6-770d-4eb8-a08b-f1727f1a58e4_guilherme-romano-KI2KaOeT670-unsplash.jpg?auto=compress,format\\">non proident</a>, sunt in culpa qui offic<span class=\\"testLabel\\">ia deserunt mollit anim id est laborum.</span></p>",
  "<img src=\\"https://images.prismic.io/200629-sms-hoy/f0a757f6-770d-4eb8-a08b-f1727f1a58e4_guilherme-romano-KI2KaOeT670-unsplash.jpg?auto=compress,format\\" alt=\\"An Atlantic Puffin\\" />",
  "<img src=\\"https://images.prismic.io/200629-sms-hoy/f0a757f6-770d-4eb8-a08b-f1727f1a58e4_guilherme-romano-KI2KaOeT670-unsplash.jpg?auto=compress,format\\" alt=\\"An Atlantic Puffin\\" />",
  "<p><span class=\\"testLabel\\">Lorem ipsum dolor <strong>sit amet</strong>, consecte</span>tur <em>adipiscing elit</em>, sed do eiusmod tempor incididunt ut labore et dolore <span class=\\"label\\">magna aliqua</span>. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea <a href=\\"https://google.com\\" target=\\"undefined\\">commodo consequat</a>. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat <a href=\\"linkResolver(XvoFFREAAM0WGBng)\\">nulla pariatur</a>. Excepteur sint occaecat cupidatat <a href=\\"https://images.prismic.io/200629-sms-hoy/f0a757f6-770d-4eb8-a08b-f1727f1a58e4_guilherme-romano-KI2KaOeT670-unsplash.jpg?auto=compress,format\\">non proident</a>, sunt in culpa qui officia deserunt mollit anim id est laborum.</p>",
  "<img src=\\"https://images.prismic.io/200629-sms-hoy/f0a757f6-770d-4eb8-a08b-f1727f1a58e4_guilherme-romano-KI2KaOeT670-unsplash.jpg?auto=compress,format\\" alt=\\"An Atlantic Puffin\\" />",
  "<p><span class=\\"label\\">Lorem ipsum dolor <strong>sit amet</strong>, consecte</span>tur <em>adipiscing elit</em>, sed do eiusmod tempor incididunt ut labore et dolore <span class=\\"label\\">magna aliqua</span>. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea <a href=\\"https://google.com\\" target=\\"undefined\\">commodo consequat</a>. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat <a href=\\"linkResolver(XvoFFREAAM0WGBng)\\">nulla pariatur</a>. Excepteur sint occaecat cupidatat <a href=\\"https://images.prismic.io/200629-sms-hoy/f0a757f6-770d-4eb8-a08b-f1727f1a58e4_guilherme-romano-KI2KaOeT670-unsplash.jpg?auto=compress,format\\">non proident</a>, sunt in culpa qui officia deserunt mollit anim id est laborum.</p>",
  "<pre>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.</pre>",
  "<p><span class=\\"label\\">Lorem ipsum dolor <strong>sit amet</strong>, consecte</span>tur <em>adipiscing elit</em>, sed do eiusmod tempor incididunt ut labore et dolore <span class=\\"label\\">magna aliqua</span>. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea <a href=\\"https://google.com\\" target=\\"undefined\\">commodo consequat</a>. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat <a href=\\"linkResolver(XvoFFREAAM0WGBng)\\">nulla pariatur</a>. Excepteur sint occaecat cupidatat <a href=\\"https://images.prismic.io/200629-sms-hoy/f0a757f6-770d-4eb8-a08b-f1727f1a58e4_guilherme-romano-KI2KaOeT670-unsplash.jpg?auto=compress,format\\">non proident</a>, sunt in culpa qui officia deserunt mollit anim id est laborum.</p>",
  "<iframe width=\\"200\\" height=\\"113\\" src=\\"https://www.youtube.com/embed/n5CwXuyNfoc?feature=oembed\\" frameborder=\\"0\\" allow=\\"accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture\\" allowfullscreen></iframe>",
  "<p><span class=\\"label\\">Lorem ipsum dolor <strong>sit amet</strong>, consecte</span>tur <em>adipiscing elit</em>, sed do eiusmod tempor incididunt ut labore et dolore <span class=\\"label\\">magna aliqua</span>. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea <a href=\\"https://google.com\\" target=\\"undefined\\">commodo consequat</a>. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat <a href=\\"linkResolver(XvoFFREAAM0WGBng)\\">nulla pariatur</a>. Excepteur sint occaecat cupidatat <a href=\\"https://images.prismic.io/200629-sms-hoy/f0a757f6-770d-4eb8-a08b-f1727f1a58e4_guilherme-romano-KI2KaOeT670-unsplash.jpg?auto=compress,format\\">non proident</a>, sunt in culpa qui officia deserunt mollit anim id est laborum.</p>",
  "<img src=\\"https://images.prismic.io/200629-sms-hoy/f0a757f6-770d-4eb8-a08b-f1727f1a58e4_guilherme-romano-KI2KaOeT670-unsplash.jpg?auto=compress,format\\" alt=\\"An Atlantic Puffin\\" />",
]
`;
